include/lo2s/perf/event.hpp

```diff
@@ -0,0 +1,291 @@
+        return scope.is_thread() || cpus_.empty() || cpus_.count(scope.as_cpu());
+    }
+
+    bool degrade_precision();
+
+    friend bool operator==(const Event& lhs, const Event& rhs)
+    {
+        return !memcmp(&lhs.attr_, &rhs.attr_, sizeof(struct perf_event_attr));
+    }
+
+    friend bool operator<(const Event& lhs, const Event& rhs)
+    {
+        return memcmp(&lhs.attr_, &rhs.attr_, sizeof(struct perf_event_attr));
+    }
+
+    friend bool operator>(const Event& lhs, const Event& rhs)
+    {
+        return memcmp(&lhs.attr_, &rhs.attr_, sizeof(struct perf_event_attr));
+    }
+
+protected:
+    void update_availability();
+    void set_common_attrs(bool enable_on_exec);
+
+    struct perf_event_attr attr_;
+    double scale_ = 1;
+    std::string unit_ = "#";
+
+    std::string name_ = "";
+    std::set<Cpu> cpus_;
+    Availability availability_ = Availability::UNAVAILABLE;
+
+    std::filesystem::path pmu_path_;
+    std::string pmu_name_;
+};
+
+/**
+ * Contains an event parsed from sysfs
+ * @note call on use_sampling_options() after creation to get a valid
+ * event, otherwise the availability will be set to UNAVAILABLE
+ */
+class SysfsEvent : public Event
+{
+public:
+    using Event::Event;
+    SysfsEvent(const std::string& ev_name, bool enable_on_exec = false);
+
+    void use_sampling_options(const bool& use_pebs, const bool& sampling, const bool& enable_cct);
+};
+
+/**
+ * Contains an opened instance of Event.
+ * Use any Event.open() method to construct an object
+ */
+class EventGuard
+{
+public:
+    EventGuard();
+    EventGuard(Event& ev, std::variant<Cpu, Thread> location, int group_fd, int cgroup_fd);
+
+    EventGuard(EventGuard&) = delete;
+    EventGuard& operator=(const EventGuard&) = delete;
+
+    EventGuard(EventGuard&& other)
+    {
+        std::swap(fd_, other.fd_);
+        std::swap(ev_, other.ev_);
+    }
+
+    EventGuard& operator=(EventGuard&& other)
+    {
+        std::swap(fd_, other.fd_);
+        std::swap(ev_, other.ev_);
+        return *this;
+    }
+
+    /**
+     * opens child as a counter of the calling (leader) event
+     */
+    EventGuard open_child(Event child, ExecutionScope location, int cgroup_fd = -1);
+
+    void enable();
+    void disable();
+
+    uint64_t get_id()
+    {
+        uint64_t id;
+        ioctl(fd_, PERF_EVENT_IOC_ID, &id);
+
+        return id;
+    }
+
+    void set_output(const EventGuard& other_ev);
+    void set_syscall_filter(const std::vector<int64_t>& filter);
+
+    int get_fd() const
+    {
+        return fd_;
+    }
+
+    bool is_valid() const
+    {
+        return fd_ >= 0;
+    };
+
+    template <class T>
+    T read()
+    {
+        static_assert(std::is_pod_v<T> == true);
+        T val;
+
+        if (::read(fd_, &val, sizeof(val)) == -1)
+        {
+            throw_errno();
+        }
+
+        return val;
+    }
+
+    ~EventGuard()
+    {
+        close(fd_);
+    }
+
+protected:
+    int fd_;
+    Event ev_;
```

Comment:
You can't make a copy of Event. In fact, I'm not sure, if you even need a reference or pointer to event at all.